src/cryptonote_basic/miner.cpp

```diff
@@ -846,40 +846,109 @@ namespace cryptonote
+            if (epee::file_io_utils::is_regular_file(power_supply_type_path))
+            {
+              std::string power_supply_type;
+              epee::file_io_utils::load_file_to_string(power_supply_type_path, power_supply_type);
+
+              if (boost::starts_with(power_supply_type, "Battery"))
+              {
+                std::string power_supply_status_path = power_supply_path + "/status";
+                if (epee::file_io_utils::is_regular_file(power_supply_status_path))
+                {
+                  std::string power_supply_status;
+                  epee::file_io_utils::load_file_to_string(power_supply_status_path, power_supply_status);
+
+                  if (boost::starts_with(power_supply_status, "Charging"))
+                  {
+                    return boost::logic::tribool(false);
+                  }
+                  else if (boost::starts_with(power_supply_status, "Full"))
+                  {
+                    return boost::logic::tribool(false);
+                  }
+                  else if (boost::starts_with(power_supply_status, "Discharging"))
+                  {
+                    return boost::logic::tribool(true);
+                  }
+                  else if (boost::starts_with(power_supply_status, "Not Charging"))
+                  {
+                    return boost::logic::tribool(boost::logic::indeterminate);
+                  }
+                  else if (boost::starts_with(power_supply_status, "Unknown"))
+                  {
+                    return boost::logic::tribool(boost::logic::indeterminate);
+                  }
+                  else
+                  {
+                    LOG_ERROR("Invalid power supply status " << power_supply_status << " for " << power_supply_path);
+                    return boost::logic::tribool(boost::logic::indeterminate);
+                  }
+                }
+                else
+                {
+                  LOG_ERROR("Unable to determine power supply status for " << power_supply_path);
+                  return boost::logic::tribool(boost::logic::indeterminate);
+                }
+              }
+              else if (boost::starts_with(power_supply_type, "Mains"))
+              {
+                std::string power_supply_online_path = power_supply_path + "/online";
+                if (epee::file_io_utils::is_regular_file(power_supply_online_path))
+                {
+                  std::string power_supply_online;
+                  epee::file_io_utils::load_file_to_string(power_supply_online_path, power_supply_online);
+
+                  if (boost::starts_with(power_supply_online, "1"))
+                  {
+                    return boost::logic::tribool(false);
+                  }
+                  else
+                  {
+                    LOG_ERROR("Unable to determine power supply status for " << power_supply_path);
+                    return boost::logic::tribool(true);
```

Comment:
I think @jtgrassie makes a good point. This should be fairly easy - start in the `true` state before the loop, remove all branches returning `true`, assign on the branches returning `indeterminate`, and then leave the `false` branches unchanged. At the end of the loop return the state.